Implement option in WCS class to preserve the original units

astrofrog opened the pull request:
### Description

A common issue users run into when using the ``astropy.wcs`` class is that WCSLIB will convert units to SI. For instance, celestial WCS in arcseconds will be converted to degrees, and spectral axes in GHz and nm will be converted to m. There isn't much we can do to stop WCSLIB from doing this - however I have realized that since we never expose the WCSLIB objects directly, but instead have our own C wrapper class, we can make that wrapper class remember any initial units and do on-the-fly conversions for all appropriate methods/properties.

As an example:

```python
In [1]: from astropy.io import fits
   ...: 

In [2]: header = fits.Header()
   ...: header['CTYPE1'] = 'RA---TAN'
   ...: header['CTYPE2'] = 'DEC--TAN'
   ...: header['CUNIT1'] = 'arcsec'
   ...: header['CUNIT2'] = 'arcsec'
   ...: header['CRVAL1'] = 20
   ...: header['CRVAL2'] = 20
   ...: header['CRPIX1'] = 1
   ...: header['CRPIX2'] = 1
   ...: header['CDELT1'] = -1
   ...: header['CDELT2'] = 1
   ...: 

In [3]: from astropy.wcs import WCS
   ...: 

In [4]: wcs = WCS(header)

In [5]: wcs.wcs.cunit
Out[5]: ['deg', 'deg']

In [6]: wcs.wcs.crval
Out[6]: array([0.00555556, 0.00555556])

In [8]: wcs.all_pix2world(1, 1, 0)
Out[8]: [array(0.00527778), array(0.00583333)]

In [9]: wcs2 = WCS(header, preserve_units=True)

In [10]: wcs2.wcs.cunit
Out[10]: ['arcsec', 'arcsec']

In [11]: wcs2.wcs.crval
Out[11]: array([20., 20.])

In [13]: wcs2.all_pix2world(1, 1, 0)
Out[13]: [array(18.99999999), array(21.)]

```

This is not yet complete but I am putting it up already so that people who are interested can try it out.

Remaining TODOs:

```
* [ ] Extend support to headers with CD matrices
* [ ] Fix/check support for printing the WCS
* [ ] Check that we can dynamically change the units on the WCS and that things behave correctly
* [ ] Expand test suite
* [ ] Write documentation
* [ ] Add news fragment

This will need to be rebased once https://github.com/astropy/astropy/pull/18231 is merged

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

astrofrog commented:
I'm seeing a test failure in ``test_wtbarr_print`` which is a test unrelated to the changes here but what is happening is that the ``print_contents()`` call there includes some output from the new test I added in this PR, so somehow something isn't getting cleared in memory.

mcara commented:
@astrofrog While I review this, may I ask whether you had contacted Mark to see if this could be fixed in WCSLIB? It seems it would be a much more natural place for the fix as it would avoid unnecessary copying of `Wcsprm` structure. It seems unnatural/unexpected for WCSLIB to (kind of) arbitrarily convert units in the `Wcsprm` structure.

Section 2.3 of Paper I says:

> 2.3. Keyword value units
The original FITS paper (Wells et al. 1981) assumed that the units along each axis could be implied simply by the contents of the CTYPEi keyword. This has not turned out to be true in general. Therefore, we propose adding a new indexed, keyword CUNITi (character-valued) with which the units of CRVALi and CDELTi may be specified. Restrictions on the nature and range of units, if any, will be determined by the agreements applying to the specific axes. If they are not so limited, units should conform with the recommendations of the IAU Style Manual (McNally 1988).

Although in Paper II (section 1) it says:
> Wells et al. (1981) established that all angles in FITS were to be measured in degrees and this has been entrenched by the AIPS convention and confirmed in the IAU-endorsed FITS standard (Hanisch et al. 2001). Paper I introduced the CUNITia keyword to define the units of CRVALia and CDELTia. **Accordingly, we require CUNITia = 'deg' for the celestial CRVALia and CDELTia, whether given explicitly or not. Consequently, the (x; y) coordinates in the plane of projection are measured in degrees.** For consistency, we use degree measure for native and celestial spherical coordinates and for all other angular measures in this paper.

This is confusing (to me). So I looked at the Hanisch et al. 2001 paper, section 5.3, and it says:

> The units of all FITS header keyword values, with the exception of measurements of angles, should conform with the recommendations in the IAU Style Manual (McNally 1988). For angular measurements given as floating point values and specified with reserved keywords, degrees are the recommended units** (with the units, if specified, given as ’deg’).

recommended != required. I think it is worth asking why WCSLIB chose to behave like that and whether it is a bug and regardless of bug or not - can't this be handled there?

While I think it is worth contacting Mark Calabretta about this, I think we can go ahead and implement this here.

mcara commented on astropy/wcs/src/wcslib_wrap.c:
```diff
@@ -1415,10 +1575,14 @@ PyWcsprm_p2s(
+  // calling wcsp2s, but we need to call it ourselves using PyWcsprm_cset so that we can
+  // catch cases where the units might change if e.g. they are not in SI to start with.
+  /* Force a call to wcsset here*/
+  if (self->preserve_units && PyWcsprm_cset(self, 1)) return NULL;
```
Even for 1-line conditional, I recommend enclosing `return NULL;` in `{}`.

astrofrog replied:
Done

mcara commented on astropy/wcs/src/wcslib_wrap.c:
```diff
@@ -1517,14 +1691,35 @@ PyWcsprm_s2p(
+  // calling wcsp2s, but we need to call it ourselves using PyWcsprm_cset so that we can
+  // catch cases where the units might change if e.g. they are not in SI to start with.
+  /* Force a call to wcsset here*/
+  if (self->preserve_units && PyWcsprm_cset(self, 1)) return NULL;
```
What is the performance cost for this for the case of a single coordinate transformation (a pair of coordinates) and for a large array of coordinates?

Again, `{ return NULL; }`

astrofrog replied:
Fixed - I'll address performance in the main discussion below

mcara commented on astropy/wcs/tests/test_wcs.py:
```diff
@@ -1915,3 +1923,349 @@ def test_DistortionLookupTable():
+        self.header = fits.Header.fromstring(HEADER_WITH_NON_SI_UNITS, sep="\n")
+        self.wcs_default = wcs.WCS(self.header)
+        self.wcs_preserve = wcs.WCS(self.header, preserve_units=True)
+        self.ones = np.ones((3, 5))
```
I think `ones` is not a good description of what this array is (for). Maybe something like `ref_coordinates` or something like that. Would it also be possible to make this array with values other than 0 or 1? Maybe make it a larger array (but !=5, of course)?

astrofrog replied:
Done - using deterministic random values now

mcara commented:
I also wonder how this works with the `-TAB` non-linear algorithm (see section 6 in Paper III - spectral). Maybe add an example using -TAB?

astrofrog commented on astropy/wcs/src/wcslib_wrap.c:
```diff
@@ -167,6 +167,50 @@ convert_rejections_to_warnings() {
   return status;
 }
 
+PyObject* get_scaled_double_array_readonly(
+    /*@unused@*/ const char* propname,
+    double* value,
+    double *scale,
+    const npy_intp naxis,
+    const npy_intp* dims,
+    /*@shared@*/ PyObject* owner) {
```
Added

astrofrog commented on astropy/wcs/src/astropy_wcs.c:
```diff
@@ -244,13 +249,19 @@ Wcs_all_pix2world(
+  // calling wcsp2s, but we need to call it ourselves using PyWcsprm_cset so that we can
+  // catch cases where the units might change if e.g. they are not in SI to start with.
+  /* Force a call to wcsset here*/
+  if (((PyWcsprm*)(self->py_wcsprm))->preserve_units && PyWcsprm_cset(((PyWcsprm*)(self->py_wcsprm)), 1)) return NULL;
```
Fixed

astrofrog commented:
> What is the performance cost for this for the case of a single coordinate transformation (a pair of coordinates) and for a large array of coordinates?

Using the 5-D WCS from the test added in this PR (``wcs_default`` is without ``preserve_units=True``, and ``wcs_preserve`` is with), and doing a lot of repeated measurements, I get:

**Before this PR**

wcs_default.all_pix2world(2, 3, 8, 1, 2, 0)        0.014 +/- 0.002 ms
wcs_default.all_pix2world(coords, 0)               1.012 +/- 0.076 ms
wcs_default.all_world2pix(2, 3, 8, 1, 2, 0)        0.028 +/- 0.008 ms
wcs_default.all_world2pix(coords, 0)               1.178 +/- 0.146 ms

**With this PR**

```
wcs_default.all_pix2world(2, 3, 8, 1, 2, 0)        0.013 +/- 0.001 ms
wcs_default.all_pix2world(coords, 0)               1.112 +/- 0.017 ms
wcs_default.all_world2pix(2, 3, 8, 1, 2, 0)        0.017 +/- 0.001 ms
wcs_default.all_world2pix(coords, 0)               1.024 +/- 0.057 ms
wcs_preserve.all_pix2world(2, 3, 8, 1, 2, 0)       0.022 +/- 0.001 ms
wcs_preserve.all_pix2world(coords, 0)              1.204 +/- 0.087 ms
wcs_preserve.all_world2pix(2, 3, 8, 1, 2, 0)       0.028 +/- 0.001 ms
wcs_preserve.all_world2pix(coords, 0)              0.221 +/- 0.013 ms
```

Time times for ``wcs_default`` are consistent within errors, so no performance impact for the default case.

With preserve units, it seems pix2world is a little slower for ``pix2world``, which is to be expected, but I am a bit surprised by ``world2pix`` actually being faster, so I will try and investigate that.

astrofrog commented:
@mcara - I am having issues getting a simple -TAB example to work. I would have thought the following would work:

```python
import numpy as np
from astropy.io import fits
from astropy.wcs import WCS

flux = np.random.rand(100)

wav = np.linspace(4000, 5000, 100)

primary_hdu = fits.PrimaryHDU(flux)

col = fits.Column(name='WAVE', format='D', array=wav)
tab_hdu = fits.BinTableHDU.from_columns([col])
tab_hdu.name = 'WCS-TAB'

header = primary_hdu.header
header['WCSAXES'] = 1
header['CTYPE1'] = 'WAVE-TAB'
header['CRPIX1'] = 1
header['CDELT1'] = 1
header['CRVAL1'] = 0
header['CUNIT1'] = 'mm'
header['PS1_0'] = 'WCS-TAB'
header['PS1_1'] = 'WAVE'
header['PV1_1'] = 1

hdul = fits.HDUList([primary_hdu, tab_hdu])

wcs_default = WCS(hdul[0].header, hdul)

print(wcs_default.wcs_pix2world(0, 0))
print(wcs_default.wcs_pix2world(1, 0))
print(wcs_default.wcs_pix2world(2, 0))
print(wcs_default.wcs_pix2world(3, 0))
```

This outputs:

```
[array(nan)]
[array(4000.)]
[array(nan)]
[array(nan)]
```

Only pixel values between 1 and 2 seem to return a valid world coordinate. Have I made a mistake in setting up my FITS file?

mcara commented:
```python
In [2]: wp = wcs_preserve.all_world2pix(world_coords_native, 0)

In [3]: wnp = wcs_default.all_world2pix(world_coords_si, 0)

In [4]: wp
Out[4]:
array([[-1.        ,         inf, -1.        , -4.        ,  2.71759582],
       [-1.        ,         inf, -1.        , -4.        ,  2.79517881],
       [-1.        ,         inf, -1.        , -4.        ,  2.82613895],
       ...,
       [-1.        ,         inf, -1.        , -4.        ,  2.73230782],
       [-1.        ,         inf, -1.        , -4.        ,  2.81918181],
       [-1.        ,         inf, -1.        , -4.        ,  2.81773117]],
      shape=(10000, 5))

In [5]: wnp
Out[5]:
array([[-0.78075305, -0.51671735, -0.55508113, -3.88556918,  2.71759582],
       [-0.83539454, -0.5772501 , -0.50062671, -3.97816059,  2.79517881],
       [-0.94703594, -0.35614207, -0.8485017 , -3.50202099,  2.82613895],
       ...,
       [-0.81628231, -0.54181075, -0.96587219, -3.10801576,  2.73230782],
       [-0.83270913, -0.58271578, -0.88742555, -3.84659959,  2.81918181],
       [-0.99437225, -0.38951446, -0.72503993, -3.45589686,  2.81773117]],
      shape=(10000, 5))

In [6]: wcs_default
Out[6]:
WCS Keywords

Number of WCS axes: 5
CTYPE : 'RA---TAN' 'FREQ' 'DEC--TAN' 'WAVE' 'OFFSET'
CRVAL : 0.0011111111111111111 5000000000.0 0.0016666666666666666 7.000000000000001e-09 8.0
CRPIX : 1.0 2.0 3.0 4.0 5.0
PC1_1 PC1_2 PC1_3 PC1_4 PC1_5  : 1.0 0.0 0.0 0.0 0.0
PC2_1 PC2_2 PC2_3 PC2_4 PC2_5  : 0.0 1.0 0.0 0.0 0.0
PC3_1 PC3_2 PC3_3 PC3_4 PC3_5  : 0.0 0.0 1.0 0.0 0.0
PC4_1 PC4_2 PC4_3 PC4_4 PC4_5  : 0.0 0.0 0.0 1.0 0.0
PC5_1 PC5_2 PC5_3 PC5_4 PC5_5  : 0.0 0.0 0.0 0.0 1.0
CDELT : 0.0011111111111111111 3000000000.0 0.0005555555555555556 1e-09 6.0
NAXIS : 0  0
```

You can see that something is not working well here. Only the `'OFFSET'` column looks good. Also, since there are no distortions, `all_world2pix` should revert to `wcs_world2pix` (after a minor overhead).

astrofrog commented:
Oh I see the issue, it looks like I am modifying the input world array in-place - oops!

mcara commented:
@pllim

```
git clone -b wcs-units-c  --single-branch https://github.com/astrofrog/astropy.git astropy-cunit
```

mcara commented:
@astrofrog So maybe add a roundtrip test (not a separate test - an `assert` would be enough) to detect these errors

mcara commented:
@astrofrog I will take a look at it a little later. Why not use https://github.com/astropy/astropy/blob/main/astropy/wcs/tests/data/tab-time-last-axis.fits (permanently or meanwhile)

mcara commented:
> Yes sounds good I can do that, but curious whether the issue I have is a real bug

me too, so I will look into it

mcara commented:
@astrofrog The following change fixes your -TAB code (don't ask why):
```python
arr = np.array([(wav, )], dtype=[("WAVE", np.float64, wav.size)])
tab_hdu = fits.BinTableHDU(arr)
```

#### Previously hdul[1].data were:
```python
FITS_rec([(4000.        ,), (4010.1010101 ,), (4020.2020202 ,),
          (4030.3030303 ,), (4040.4040404 ,), (4050.50505051,),
          (4060.60606061,), (4070.70707071,), (4080.80808081,),
          (4090.90909091,), (4101.01010101,), (4111.11111111,),
          (4121.21212121,), (4131.31313131,), (4141.41414141,),
          (4151.51515152,), (4161.61616162,), (4171.71717172,),
          (4181.81818182,), (4191.91919192,), (4202.02020202,),
          (4212.12121212,), (4222.22222222,), (4232.32323232,),
          (4242.42424242,), (4252.52525253,), (4262.62626263,),
          (4272.72727273,), (4282.82828283,), (4292.92929293,),
          (4303.03030303,), (4313.13131313,), (4323.23232323,),
          (4333.33333333,), (4343.43434343,), (4353.53535354,),
          (4363.63636364,), (4373.73737374,), (4383.83838384,),
          (4393.93939394,), (4404.04040404,), (4414.14141414,),
          (4424.24242424,), (4434.34343434,), (4444.44444444,),
          (4454.54545455,), (4464.64646465,), (4474.74747475,),
          (4484.84848485,), (4494.94949495,), (4505.05050505,),
          (4515.15151515,), (4525.25252525,), (4535.35353535,),
          (4545.45454545,), (4555.55555556,), (4565.65656566,),
          (4575.75757576,), (4585.85858586,), (4595.95959596,),
          (4606.06060606,), (4616.16161616,), (4626.26262626,),
          (4636.36363636,), (4646.46464646,), (4656.56565657,),
          (4666.66666667,), (4676.76767677,), (4686.86868687,),
          (4696.96969697,), (4707.07070707,), (4717.17171717,),
          (4727.27272727,), (4737.37373737,), (4747.47474747,),
          (4757.57575758,), (4767.67676768,), (4777.77777778,),
          (4787.87878788,), (4797.97979798,), (4808.08080808,),
          (4818.18181818,), (4828.28282828,), (4838.38383838,),
          (4848.48484848,), (4858.58585859,), (4868.68686869,),
          (4878.78787879,), (4888.88888889,), (4898.98989899,),
          (4909.09090909,), (4919.19191919,), (4929.29292929,),
          (4939.39393939,), (4949.49494949,), (4959.5959596 ,),
          (4969.6969697 ,), (4979.7979798 ,), (4989.8989899 ,),
          (5000.        ,)],
         dtype=(numpy.record, [('WAVE', '<f8')]))
```

#### Now hdul[1].data is:
```python
FITS_rec([([4000.        , 4010.1010101 , 4020.2020202 , 4030.3030303 , 4040.4040404 , 4050.50505051, 4060.60606061, 4070.70707071, 4080.80808081, 4090.90909091, 4101.01010101, 4111.11111111, 4121.21212121, 4131.31313131, 4141.41414141, 4151.51515152, 4161.61616162, 4171.71717172, 4181.81818182, 4191.91919192, 4202.02020202, 4212.12121212, 4222.22222222, 4232.32323232, 4242.42424242, 4252.52525253, 4262.62626263, 4272.72727273, 4282.82828283, 4292.92929293, 4303.03030303, 4313.13131313, 4323.23232323, 4333.33333333, 4343.43434343, 4353.53535354, 4363.63636364, 4373.73737374, 4383.83838384, 4393.93939394, 4404.04040404, 4414.14141414, 4424.24242424, 4434.34343434, 4444.44444444, 4454.54545455, 4464.64646465, 4474.74747475, 4484.84848485, 4494.94949495, 4505.05050505, 4515.15151515, 4525.25252525, 4535.35353535, 4545.45454545, 4555.55555556, 4565.65656566, 4575.75757576, 4585.85858586, 4595.95959596, 4606.06060606, 4616.16161616, 4626.26262626, 4636.36363636, 4646.46464646, 4656.56565657, 4666.66666667, 4676.76767677, 4686.86868687, 4696.96969697, 4707.07070707, 4717.17171717, 4727.27272727, 4737.37373737, 4747.47474747, 4757.57575758, 4767.67676768, 4777.77777778, 4787.87878788, 4797.97979798, 4808.08080808, 4818.18181818, 4828.28282828, 4838.38383838, 4848.48484848, 4858.58585859, 4868.68686869, 4878.78787879, 4888.88888889, 4898.98989899, 4909.09090909, 4919.19191919, 4929.29292929, 4939.39393939, 4949.49494949, 4959.5959596 , 4969.6969697 , 4979.7979798 , 4989.8989899 , 5000.        ],)],
         dtype=(numpy.record, [('WAVE', '<f8', (100,))]))
```

mcara commented:
I think the coordinate array is an element of the BinTable

mcara commented:
I remember it is standard

mcara commented:
#### Old header:
```
XTENSION= 'BINTABLE'           / binary table extension
BITPIX  =                    8 / array data type
NAXIS   =                    2 / number of array dimensions
NAXIS1  =                    8 / length of dimension 1
NAXIS2  =                  100 / length of dimension 2
PCOUNT  =                    0 / number of group parameters
GCOUNT  =                    1 / number of groups
TFIELDS =                    1 / number of table fields
TTYPE1  = 'WAVE    '
TFORM1  = 'D       '
EXTNAME = 'WCS-TAB '           / extension name
```
#### New Header:
```
XTENSION= 'BINTABLE'           / binary table extension
BITPIX  =                    8 / array data type
NAXIS   =                    2 / number of array dimensions
NAXIS1  =                  800 / length of dimension 1
NAXIS2  =                    1 / length of dimension 2
PCOUNT  =                    0 / number of group parameters
GCOUNT  =                    1 / number of groups
TFIELDS =                    1 / number of table fields
TTYPE1  = 'WAVE    '
TFORM1  = '100D    '
TDIM1   = '(100)   '
EXTNAME = 'WCS-TAB '           / extension name
```

#### From Paper III
> 6.1.3. -TAB implementation, parameters, and requirements

>Standard FITS binary tables extensions (XTENSION =
‘BINTABLE’) will be used to hold the coordinate array in
a single cell of a column of a one-row table. The length of
this array is given in the FITS table header by the repeat
count in the TFORMn keyword, where n is the number of the
column containing the coordinate array. The dimensions of the
coordinate array will be given in the FITS table header by the
keyword TDIMn set to ‘(M, K1, K2, . . . KM)’, where n is
the column containing the coordinate array. Note in particular
that if M = 1 the coordinate array is formally two-dimensional
though the first axis is degenerate. The repeat count in the
TFORMn keyword value is the product of M and all the Km.

astrofrog commented:
Thanks! I've fixed the test

astrofrog commented:
Ok I consider this to be fully ready for review/testing (as I've added the docs)

I will try and do some real world testing of other packages with the ``preserve_units=True`` option tomorrow.

mcara commented on astropy/wcs/tests/test_wcs.py:
```diff
@@ -1934,3 +1943,412 @@ def test_DistortionLookupTable():
+            [x.rstrip() for x in strip_memory_addresses(captured.out).splitlines()]
+        )
+
+    def test_tab(self):
```
Please change the test name to something more detailed: `test_wcstab_cunit_conversion` or similar. Currently it is too generic, i.e., suggesting it is testing that support `-TAB` WCS in general works.

astrofrog replied:
Just to be clear, the full name of the test is ``TestPreserveUnits.test_tab``. However, I have renamed it anyway following your suggestion.

mcara commented:
```
In [62]: wcs_preserved.wcs.cunit
Out[62]: ['mm']

In [63]: wcs_default.wcs.cunit
Out[63]: ['mm']
```
It looks like in the case of TAB WCS, WCSLIB is not doing unit conversion. Can you print what is the reported scale/conversion factor for this case? So, first, I wish the unit test would detect this but it is not the most important thing. The inconsistency in WCSLIB in how it treats WCS-TAB vs regular WCS is what I don't like. Is this a problem or not? If the behavior of WCSLIB with regard to unit conversion in -TAB WCS changes in the future then it would be nice for the unit test to detect that. However, currently it means that we can disable this conversion (unit preservation) for -TAB for the performance reasons: there is no conversion going on here. What is your opinion?

Cadair commented:
> It looks like in the case of TAB WCS, WCSLIB is not doing unit conversion.

Yes, this is true. It's caused other bugs in the past. I think we should just leave it.

astrofrog commented:
> It looks like in the case of TAB WCS, WCSLIB is not doing unit conversion. Can you print what is the reported scale/conversion factor for this case? So, first, I wish the unit test would detect this but it is not the most important thing. The inconsistency in WCSLIB in how it treats WCS-TAB vs regular WCS is what I don't like. Is this a problem or not? If the behavior of WCSLIB with regard to unit conversion in -TAB WCS changes in the future then it would be nice for the unit test to detect that. However, currently it means that we can disable this conversion (unit preservation) for -TAB for the performance reasons: there is no conversion going on here. What is your opinion?

I do find it annoying how sometimes WCSLIB will convert and sometimes not, though this is beyond the scope of this PR.

I have made several changes/additions since you last looked at this:

* There is now a private property ``WCS.wcs._unit_scaling`` which gives an array with the scalefactors being used. This is private because I don't think users should have access to it, it is just for us to debug/understand how things are working and we can use it in tests. This property returns ``None`` if there is no scaling going on
* If ``preserve_units=False``, ``_unit_scaling`` is always ``None``
* If ``preserve_units=True``, then ``_unit_scaling`` is only set if at least one of the scalefactors is not 1 - this addresses your question about TAB - in this case, ``_unit_scaling`` is just ``None``. This also covers cases where the unit is fixed by WCSLIB but with no change in scalefactor, e.g. ``'deg '`` -> ``'deg'``
* I've added checks for ``_unit_scaling`` in a few of the tests to make sure things are working as we expect
* I've added an explicit check for the units in the TAB test so that we can detect any change in the default WCSLIB behavior in future

Does this all make sense?

mcara reviewed (approved):
This looks good. Many thanks!

mcara commented on astropy/wcs/src/wcslib_wrap.c:
```diff
@@ -3997,6 +4348,24 @@ PyWcsprm_set_zsource(
 
 
+ /*@null@*/ static PyObject*
+PyWcsprm_get_unit_scaling(
+    PyWcsprm* self,
+    /*@unused@*/ void* closure) {
```
Maybe add some documentation (for the developers) here: when does it return None and when an array of let's say 1s? What is the difference between `None` and `[1, 1]`?

mcara replied:
Would it be more consistent to return a list of `[None, None]` instead of `[1, 1]` or when only some axes need conversion then `[None, None, 1e3]`?

astrofrog replied:
Given that this is private API I would rather keep it simple and just mirror what is happening in the C code since it's just to examine unit_scaling.

Either we use the unit conversion path in the code in which case we have an array of floats or we don't use the conversions in which case we return None. We couldn't easily return a mixed double/None array without using an object array or a list, and it's not what's happening in the C code.

Would you be happy for me to leave as is if I just add a docstring?

mcara commented:
Yes, there is something weird going on here. The following commands are run after your example from https://github.com/astropy/astropy/pull/18338#issuecomment-3073605858 (with the fixed array stuff) is run as first input.

### Case 1: preserve_units=False

```python
In [2]: with fits.open('astropy/astropy/wcs/tests/data/tab-time-last-axis.fits') as h:
   ...:     w0 = WCS(h[0].header, h, preserve_units=False)
   ...:     print(w0)
   ...:     print("\n\n----- NON-SI time -----")
   ...:     # make TIME axis with non-SI units:
   ...:     h[0].header["CUNIT3"] = "min"
   ...:     w1 = WCS(h[0].header, h, preserve_units=False)
   ...:     print(w1)
   ...:
WCS Keywords

Number of WCS axes: 3
CTYPE : 'RA---TAN' 'DEC--TAN' 'TIME-TAB'
CUNIT : 'deg' 'deg' 's'
CRVAL : 5.6305039831305 -72.050216354325 1.0
CRPIX : 32.5 16.5 1.0
PC1_1 PC1_2 PC1_3  : 5.9533058133115e-06 1.2909957973842e-05 0.0
PC2_1 PC2_2 PC2_3  : -1.2644950182756e-05 5.0222580309289e-06 0.0
PC3_1 PC3_2 PC3_3  : 0.0 0.0 1.0
CDELT : 1.0 1.0 1.0
NAXIS : 1  1  1


----- NON-SI time -----
WCS Keywords

Number of WCS axes: 3
CTYPE : 'RA---TAN' 'DEC--TAN' 'TIME-TAB'
CUNIT : 'deg' 'deg' 'min'
CRVAL : 5.6305039831305 -72.050216354325 1.0
CRPIX : 32.5 16.5 1.0
PC1_1 PC1_2 PC1_3  : 5.9533058133115e-06 1.2909957973842e-05 0.0
PC2_1 PC2_2 PC2_3  : -1.2644950182756e-05 5.0222580309289e-06 0.0
PC3_1 PC3_2 PC3_3  : 0.0 0.0 1.0
CDELT : 1.0 1.0 1.0
NAXIS : 1  1  1

In [3]: w0.wcs_pix2world([[1, 2, 3]], 0)
Out[3]: array([[  5.6293493 , -72.04989848,   3.        ]])

In [4]: w1.wcs_pix2world([[1, 2, 3]], 0)
Out[4]: array([[  5.6293493 , -72.04989848,   3.        ]])

In [5]: w1.wcs.cunit = ["deg", "deg", "s"]

In [6]: w1.wcs.set()

In [7]: w1.wcs_pix2world([[1, 2, 3]], 0)
Out[7]: array([[  5.6293493 , -72.04989848,   3.        ]])

In [8]: w1.wcs.cunit = ["deg", "deg", "min"]

In [9]: w1.wcs.set()

In [10]: w1.wcs_pix2world([[1, 2, 3]], 0)
Out[10]: array([[  5.6293493 , -72.04989848,   3.        ]])

In [11]: w1
Out[11]:
WCS Keywords

Number of WCS axes: 3
CTYPE : 'RA---TAN' 'DEC--TAN' 'TIME-TAB'
CUNIT : 'deg' 'deg' 'min'
CRVAL : 5.6305039831305 -72.050216354325 1.0
CRPIX : 32.5 16.5 1.0
PC1_1 PC1_2 PC1_3  : 5.9533058133115e-06 1.2909957973842e-05 0.0
PC2_1 PC2_2 PC2_3  : -1.2644950182756e-05 5.0222580309289e-06 0.0
PC3_1 PC3_2 PC3_3  : 0.0 0.0 1.0
CDELT : 1.0 1.0 1.0
NAXIS : 1  1  1

In [12]: w1.wcs.cunit = ["deg", "deg", "min"]

In [13]: w1.wcs.cunit = ["deg", "deg", "min"]

In [14]: w1.wcs.set()

In [15]: w1.wcs.cunit = ["deg", "deg", "min"]

In [16]: w1.wcs.set()
```

### Case 2: preserve_units=True

I thought TAB is not converted...

```python
In [2]: with fits.open('astropy/astropy/wcs/tests/data/tab-time-last-axis.fits') as h:
   ...:     w0 = WCS(h[0].header, h, preserve_units=True)
   ...:     print(w0)
   ...:     print("\n\n----- NON-SI time -----")
   ...:     # make TIME axis with non-SI units:
   ...:     h[0].header["CUNIT3"] = "min"
   ...:     w1 = WCS(h[0].header, h, preserve_units=True)
   ...:     print(w1)
   ...:
WCS Keywords

Number of WCS axes: 3
CTYPE : 'RA---TAN' 'DEC--TAN' 'TIME-TAB'
CUNIT : 'deg' 'deg' 's'
CRVAL : 5.6305039831305 -72.050216354325 1.0
CRPIX : 32.5 16.5 1.0
PC1_1 PC1_2 PC1_3  : 5.9533058133115e-06 1.2909957973842e-05 0.0
PC2_1 PC2_2 PC2_3  : -1.2644950182756e-05 5.0222580309289e-06 0.0
PC3_1 PC3_2 PC3_3  : 0.0 0.0 1.0
CDELT : 1.0 1.0 1.0
NAXIS : 1  1  1


----- NON-SI time -----
WCS Keywords

Number of WCS axes: 3
CTYPE : 'RA---TAN' 'DEC--TAN' 'TIME-TAB'
CUNIT : 'deg' 'deg' 'min'
CRVAL : 5.6305039831305 -72.050216354325 1.0
CRPIX : 32.5 16.5 1.0
PC1_1 PC1_2 PC1_3  : 5.9533058133115e-06 1.2909957973842e-05 0.0
PC2_1 PC2_2 PC2_3  : -1.2644950182756e-05 5.0222580309289e-06 0.0
PC3_1 PC3_2 PC3_3  : 0.0 0.0 1.0
CDELT : 1.0 1.0 1.0
NAXIS : 1  1  1

In [3]: w0.wcs_pix2world([[1, 2, 3]], 0)
Out[3]: array([[  5.6293493 , -72.04989848,   3.        ]])

In [4]: w1.wcs_pix2world([[1, 2, 3]], 0)
Out[4]: array([[  5.6293493 , -72.04989848,   3.        ]])

In [5]: w1.wcs.cunit = ["deg", "deg", "s"]

In [6]: w1.wcs.set()

In [7]: w1
Out[7]:
WCS Keywords

Number of WCS axes: 3
CTYPE : 'RA---TAN' 'DEC--TAN' 'TIME-TAB'
CUNIT : 'deg' 'deg' 'min'
CRVAL : 5.6305039831305 -72.050216354325 0.016666666666666666
CRPIX : 32.5 16.5 1.0
PC1_1 PC1_2 PC1_3  : 5.9533058133115e-06 1.2909957973842e-05 0.0
PC2_1 PC2_2 PC2_3  : -1.2644950182756e-05 5.0222580309289e-06 0.0
PC3_1 PC3_2 PC3_3  : 0.0 0.0 1.0
CDELT : 1.0 1.0 0.016666666666666666
NAXIS : 1  1  1

In [8]: w1.wcs_pix2world([[1, 2, 3]], 0)
Out[8]: array([[ 5.62934930e+00, -7.20498985e+01,  5.00000000e-02]])

In [9]: w1.wcs.cunit = ["deg", "deg", "min"]
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[9], line 1
----> 1 w1.wcs.cunit = ["deg", "deg", "min"]

AttributeError: Original units have already been set, cannot change them
```

Notice how `CDELT3` has changed and the time world coordinate too and also the crash in the end.

Something is not right here.

astrofrog commented:
And indeed in the WCSLIB code that modifies CDELT when units are changed, tabular axes are skipped (start of the following block):

https://github.com/astropy/astropy/blob/dfcc608a9fe441abccba2744062df8cbd0792b7e/cextern/wcslib/C/wcs.c#L3310-L3339

mcara commented:
> The reason an error is raised is because it's too ambiguous what it means to change the units in the context of preserve_units=True and would rather be safe and disallow it.

This means that a WCS object created from a header is "special" in some regards compared to an object created programmatically - it behaves differently. I really do not like that. I think if user (caller) wants to change the units - they should be able to do that and the object should behave like it used to be before (before your PR). I think this could be achieved by disabling `preserve_units=True` setting (i.e., set it to `False`) when `cunit` is programmatically set or modified. This is easy to do in `PyWcsprm_set_cunit` when `cunit` is set to a list, like
```python
wcs_obj.wcs.cunit = ["deg", "deg", "m"]
```
but it is much more difficult when it is modified in place:
```python
wcs_obj.wcs.cunit[2] = "m"
```
(I think it could be done in the `Wcsprm.set()` by comparing old and new values but it should be done after the first call to "set" - that is trickier).

mcara commented:
I just run the example for `preserve_units=True` and it behaves at it should: it allowed me to modify the units, that is: no difference in behavior between an object constructed with `preserve_units=True` and `preserve_units=False`

mcara commented:
Let me review everything again.

mcara commented:
> @mcara - you said:
> 
> > I thought TAB is not converted...
> 
> If I look at the output from your examples, indeed even in the `preserve_units=False` case, the `min` units are preserved, so this is what I meant.

What _I_ was referring to was the change in CDELT3 (for the TAB axis) value in my example:
```
CDELT : 1.0 1.0 1.0
In [5]: w1.wcs.cunit = ["deg", "deg", "s"]
In [6]: w1.wcs.set()
CDELT : 1.0 1.0 0.016666666666666666
```
That is a "conversion" that should not be happening.

> For the error you get at the end:
> 
> ```
> AttributeError: Original units have already been set, cannot change them
> ```
> 
> this is good as I've deliberately made it so that the units can't be changed once they have been set - however you should also not have been able to set them the first time when you did `w1.wcs.cunit = ["deg", "deg", "s"]`, so I am going to investigate why this is. The reason an error is raised is because it's too ambiguous what it means to change the units in the context of `preserve_units=True` and would rather be safe and disallow it.

Based on my last re-test with my example, it looks like you are no longer placing this restriction, is that correct?

mcara commented:
Well, I re-run again that example and now it will not let me set the units programmatically. I must have not re-started the interpreter.

mhvk reviewed (commented):
@astrofrog - just looked over this since it involved units. Overall, of course great to do this. Two more general comments (sorry if this is duplication; I did not read the existing discussion...):
- In the implementation, the cases with and without conversion are consistently split. I wonder if it isn't easier to just always convert, but have conversion factor 1 if no action is to be taken (and similarly override `original_unit` with whatever wcslib decides it should be). Then, only in the true performance critical sections, one could just check if the conversion factor equals 1 and skip as needed. (Since this is essentially an implementation detail, it can of course be done later.)
- From the python side, doing this consistently with an argument seems painful. Could one let the default depend on a configuration item? (Again, this can probably be done in follow-up).

mcara commented:
Not identically the same but a similar underlying issue: https://github.com/astropy/astropy/issues/11215#issuecomment-755906300

I have a feeling this may have been brought up somewhere else as well. It's not impossible to deal with this but, IMO, not worth the effort.

astrofrog commented:
Ok sounds good thanks for tracking this down! I will try and wrap this up early next week.

astrofrog commented:
@mcara - I spent some time today working on making sure we can assign units to ``WCS.wcs.cunit``, and I have something that seems to work. I have added a few tests for different scenarios. I've also made it so that if ``preserve_units=True``, and ``WCS.wcs.set()`` has been called, users can't modify individual items of ``WCS.wcs.cunit[1] in-place since that is difficult to deal with otherwise, and better to just have an explicit error. This won't break any existing usage since it only applies to when ``preserve_units=True``.

In any case, I was wondering if you could take this for another spin and see if you are happy with the behavior under different circumstances?

astrofrog commented:
@mhvk thank you for taking a look at this! Regarding the implementation i would prefer for now to keep the code paths reasonably separate for preserve_units False or True - mainly to avoid introducing bugs for regular users of the default WCS class - but we could always try and tighten up the implementation once it's had a good amount of real world testing?

I would like to try and add a ScienceState to control the default behavior, but would prefer to keep this for a separate PR.

mcara reviewed (approved):
Masterclass

pllim commented on docs/changes/wcs/18338.feature.rst:
Oooo... does this also simplify unit handling for Jdaviz downstream? cc @kecnry

Cadair commented:
Thanks for doing this @astrofrog this is something which has been a pain point for a long time.

pllim commented:
I dare say this might be a killer feature for the next release. 😆 😭

Cadair commented:
I'm sure if we turned it on by default *nothing* would break.

mcara commented:
I think it would be important to have a method/property on the WCS object like `is_preserving_units` for downstream code to test which behavior is the object using. I forgot to mention this when reviewing this. Would it be possible to add this in a follow-up PR, @astrofrog ?

astrofrog commented:
@mcara - just to check, would it be better to simply have a ``WCS.preserve_units`` property? The reason I ask is because often a pattern is that things passed as kwargs to classes can then be accessed as properties, so there is some symmetry:

```
>>> wcs = WCS(..., preserve_units=True)
>>> wcs.preserve_units
True
```

Or would you prefer a method for this?

@mhvk @Cadair if you have opinions feel free to chime in!

Cadair commented:
+1 property which can't be set.

mcara commented:
I like property

mcara commented:
For the record, `drizzlepac` regression tests all passed: https://github.com/spacetelescope/RegressionTests/actions/runs/16503552056

mcara commented:
> With the default behaviour or opting into preserve units? 😉

Default as `drizzlepac` is not yet aware of `preserve_units` and likely it won't ever use this since, I believe, HST WCS for all instruments use SI (but I may be wrong).

mcara commented:
I think a global switch is bound to have consequences on other packages. For example, a package that is expecting RA & DEC to be in degrees will not be aware that, due to a global switch, it now gets angles in arcsec (for example).
